openstreetmap-carto.lua

```diff
@@ -0,0 +1,518 @@
+                for i = 1, membercount do
+                    members_superseeded[i] = 1
+                end
+                combined_tags.z_order = z_order(keyvalues)
+                return 0, combined_tags, members_superseeded, 0, 1, 0
+            end
+        else
+            -- This is a new-style MP
+            keyvalues.z_order = z_order(keyvalues)
+            return 0, keyvalues, members_superseeded, 0, 1, 0
+        end
+        assert(false, "End of control reached prematurely for MP")
+    elseif type == "route" then
+        if next(keyvalues) ~= nil then
+            keyvalues.z_order = z_order(keyvalues)
+            return 0, keyvalues, members_superseeded, 1, 0, roads(keyvalues)
+        end
+    end
+
+    -- Untagged or unknown type
+    return 1, keyvalues, members_superseeded, 0, 0, 0
+end
+
+--- Check if an object with given tags should be treated as polygon
+-- @param tags OSM tags
+-- @return 1 if area, 0 if linear
+function isarea (tags)
+    -- Treat objects tagged as area=yes polygon, other area as no
+    if tags["area"] then
+        return tags["area"] == "yes" and 1 or 0
+    end
+
+   -- Search through object's tags
+    for k, v in pairs(tags) do
+        -- Check if it has a polygon key and not a linestring override, or a polygon k=v
+        for _, ptag in ipairs(polygon_keys) do
+            if k == ptag and not (linestring_values[k] and linestring_values[k][v]) then
+                return 1
+            end
+        end
+
+        if (polygon_values[k] and polygon_values[k][v]) then
+            return 1
+        end
+    end
+    return 0
+end
+
+function is_in (needle, haystack)
+    for index, value in ipairs (haystack) do
+        if value == needle then
+            return true
+        end
+    end
+    return false
+end
+
+--- compare two values.
+-- if they are tables, then compare their keys and fields recursively.
+-- @within Comparing
+-- @param t1 A value
+-- @param t2 A value
+-- @return true or false
+function equaltables (t1,t2)
+    for k, v in pairs(t1) do
+        if t2[k] ~= v then return false end
+    end
+    for k, v in pairs(t2) do
+        if t1[k] ~= v then return false end
+    end
+    return true
+end
+
+--- Combines the tags of relation members
+-- If the tags are conflicting then nil is returned. Members with no tags are ignored
+-- @param member_tags OSM tags of relation members
+-- @return combined tags, or nil if cannot combine
+function combine_member_tags (member_tags)
+    local combined_tags = {}
+    for _, tags in ipairs(member_tags) do
+        -- Check if the member has tags
+        if next(tags) ~= nil then
+            if next(combined_tags) == nil then
+                -- This is the first tagged member
+                combined_tags = tags
+            else
+                -- A different tagged member
+                if not equaltables(tags, combined_tags) then
+                    return nil
```

Comment:
I think there's a difference between a return of {} and nil in what the function has concluded, but I see there's nothing documented. I think I have some more documentation in CT which I need to copy over, but there is a difference between an invalid MP and a valid one with no tags, but since we don't need untagged objects, the difference doesn't matter.